keps/sig-api-machinery/3617-fine-grained-authz/README.md

```diff
@@ -0,0 +1,1122 @@
+
+Currently apiserver performs the authz check before descending into any detailed
+logic for mutating operations. This means that when it is doing the authz check,
+it knows the *object* that is being changed, but nothing about *which fields*
+are changing.
+
+We will introduce a new permission, "granular". This permission
+means that although the caller doesn't have generic update permissions, it may
+still change some fields if it has the specific permissions required.
+
+API Server already maintains schema information for all resource types. This
+will be watched to also maintain field <-> permission mappings.
+
+Now we can describe the modifications to API Server:
+
+On any mutating call,
+
+1. Check the "PUT" / "PATCH" / "CREATE" permission. If the actor has this,
+   place a marker=regular in the request context and proceed to step 3.
+2. Otherwise, check the "granular" permission. If yes, place a
+   marker=granular in the request context and proceed to step 3, otherwise, fail
+   the request with a forbidden error.
+3. Compute the change (patch logic, SSA logic, defaulting etc). Compute a list
+   of fields which changed. (SSA logic makes this easy.)
+4. Pre- or post- (DECISION NEEDED) webhooks, check the list of fields:
+5. If the marker=granular, every field needs to be covered by some permission.
+   Check permissions, stop and fail if some field is not covered. (The order is
+   described below.)
+6. If the marker=regular ONLY fields having "excluded" permissions need to be
+   checked; if such fields are modified, check their permissions as in step 5.
+7. (If the marker is absent or has some other value, fail with a 5xx error.)
+8. If all checks pass, perform the rest of the needed operation.
+9. If any check fails, fail the request. The error message will not list ALL
+   permissions that are lacked, because it is possible to craft a request that
+   requires a large number of authz requests.
+
+When fine-grained field permissions need to be checked, they will be checked
+in the order of most general to most specific. When we describe the
+representation in the schema it will be clear what that means. The reason for
+this is that multiple fields can be covered by a more general permission,
+greatly reducing the number of checks needed in the worst case (as long as the
+system administrator has made use of the general permission).
+
+### Notes/Constraints/Caveats (Optional)
+
+This is a complex design. See the alternatives for why we propose it anyway.
+
+### Risks and Mitigations
+
+This design avoids the "grant an overly broad permission and then restrict it
+later" pattern, which has undesirable failure / misconfiguration patterns.
+
+This design is written to have no impact on existing cluster users for existing
+fields.
+
+A risk is that it would be possible to configure permissions which result in a
+large number of authz checks. Specifically CRD authors could craft a CRD with
+many checks required to do anything.
```

Comment:
We [restrict resource use by CRDs' validation functions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#resource-use-by-validation-functions). Perhaps a similar restriction could apply here?

Reply:
Oops, feedback is too early. Sorry about that.